_adr/92/index.adoc

```diff
@@ -0,0 +1,102 @@
+
+* the current services are offering the bare functionality leaving room for integration code to live in the services' consumer codebases
+* the data aggregation logic is duplicated, scattered, and eventually diverging in various codebases (especially UI and CLI)
+* there is no "single source of truth" for data aggregated from various services
+* the current REST API layer exposes the user to the entire complexity of the underlying infrastructure
+
+We want to deduplicate the development effort and provide a more cohesive UX to the users.
+
+## Goals
+
+* a more cohesive UX experience across UI and CLI
+* flow simplification for the REST API exposing opinionated endpoints directly mapping UI/CLI needs
+* a single place with shared logic for aggregating and serving data
+* reduce the interactions over the network for providing a given functionality
+* easy testing/mocking setup for UI/CLI
+
+## Non-goals
+
+* provide a generic API layer for consuming the current services APIs
+* provide a generic queryable layer for the current services APIs
+
+## Current situation
+
+In the current architecture both UI and CLI are reaching out both Control Plane and Data Plane REST API endpoints.
+
+image::current_architecture.png[Current Architecture]
+
+
+## Proposal
+
+[NOTE]
+*Assumption* as a first iteration we are assuming that the CLI needs can be covered by a strict subset of the functionalities needed for the UI.
+
+We propose to integrate *one* server side component commonly known as https://samnewman.io/patterns/architectural/bff/[BFF] (Backend For Frontend).
+
+image::with_BFF.png[Proposed Architecture]
+
+The service is implemented using Java and Quarkus to achieve maximum collaboration across the current teams and align with Red Hat tech strategy.
+The API layer is designed using API Designer(RHOAD) and conforming to the OpenAPI 3+ specification to achieve maximum collaboration across different teams and members with various technological background.
+The development is driven by a "Contract First" approach leveraging the [Apicurio Codegen Quarkus Extension](https://github.com/Apicurio/apicurio-codegen/tree/main/quarkus-extension) and re-using the Java SDKs(based on Kiota) provided by the DevExp Team.
+The BFF have to to be stateless so that it can easily be scaled up and down according to the needs.
+All of the exposed endpoints are going to perform 1+ REST APIs calls to the underlying services and provide back the result.
+
+### Threat model
+
+Being it a publicly accessible service the generic service Threat model applies, we should ensure stability of the platform by having all the good practices in place:
+
+* Rate Limiting
+* Circuit Breaking
+* Caching
+
+Regarding multitenancy, it can be achieved, in the first prototype by re-using the Access Token for the initial API request to issue the following requests to the backend services.
```

Comment:
In general this is something we want to facilitate, though sometimes the right thing to do is make a new request with a client credential grant (i.e. a service account authorization grant as opposed to the user's grant). It depends whether the request is really on behalf of the user, or whether it needs to be done with the identity of the BFF layer itself. This is a more complex issue of semantics and authorization which I'll leave out of this reply.

As far as making a subsequent request on behalf of the user, with a single SSO (a single issuer), yes, this becomes easier. But, a compatible issuer is not the only requirement. You also need to make sure the access token has the right audience and scope for downstream APIs. For audience ("aud" claim), we're simplifying this for new clients to be simply "redhat.com" for any hosted redhat.com APIs. Then, each API (aka "resource server) needs to define at least one scope (and ideally several) and enforce it. Then, the clients (frontend apps I believe in this case), need to be sure to request those scopes during their grants. On the SSO side, we can ensure that scope inclusions are transitive, so the client doesn't need to know about downstream calls. That is, the client might request a scope for the BFF to do something, and if that needs another service, which needs another service, those downstream scopes can be included automatically by SSO if previously negotiated between SSO provider (us) and those API providers when configuring the scopes.

See https://docs.google.com/document/d/1fv2LNl81avyimUGoHiVJzAKX3I-gZRu9cpiS0qDmz9Y/edit#heading=h.ds8q4xtkmu64 for more info.

Reply:
One minor issue to consider with this, is if we ever get to using [sender-constrained tokens](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop) or [signed HTTP requests](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures), it becomes a bit more difficult to forward tokens along. But, this is probably worth writing off as a known challenge for future "us" to worry about ;-).

Reply:
In the context of BFF, the capture is that we should prefer that we can propagate the access token throguh to the dependent services because it will keep the development of BFF easier and keep the security model similar to a usage without BFF.

Reply:
I added the detail in a subsequent commit.